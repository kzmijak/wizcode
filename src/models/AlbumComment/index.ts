export type { AlbumCommentModel } from "./AlbumCommentModel";
