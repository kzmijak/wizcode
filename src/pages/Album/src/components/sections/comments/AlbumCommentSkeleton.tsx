import { Avatar, Box, Stack, styled } from "@mui/joy";
import { FC } from "react";

const Paragraph = styled(Box)({
  backgroundColor: "lightgray",
  height: "1em",
  minWidth: 70,
});

export const AlbumCommentSkeleton: FC = () => {
  return (
    <Stack
      data-testid="comment-skeleton"
      component="article"
      direction="row"
      spacing={3}
      padding={2}
      borderRadius={10}
      border={(theme) => `dotted 2px ${theme.palette.divider}`}
    >
      <Stack spacing={2} width={60} justifyContent="center" alignItems="center">
        <Avatar variant="solid" color="neutral" />
        <Paragraph />
      </Stack>
      <Stack flex={1} spacing={2}>
        <Paragraph />
        <Paragraph minHeight="3em" />
      </Stack>
      <Paragraph />
    </Stack>
  );
};
