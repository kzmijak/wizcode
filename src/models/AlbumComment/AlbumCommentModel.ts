import { AlbumModel } from "models/Album";

export type AlbumCommentModel = {
  id: string;
  albumId: AlbumModel["id"];
  userName: string;
  userAvatar: string;
  title: string;
  description: string;
  lastModificationDate: Date;
};
