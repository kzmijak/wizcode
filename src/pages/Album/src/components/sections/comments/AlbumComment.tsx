import { Avatar, Stack, Typography } from "@mui/joy";
import { AlbumCommentModel } from "models/AlbumComment";
import { FC } from "react";

export type AlbumCommentProps = Pick<
  AlbumCommentModel,
  "description" | "lastModificationDate" | "title" | "userAvatar" | "userName"
> & {
  orientation?: "row" | "column";
};

export const AlbumComment: FC<AlbumCommentProps> = ({
  description,
  lastModificationDate,
  title,
  userAvatar,
  userName,
  orientation,
}) => {
  const dateComponent = (
    <Typography level="body2">
      {lastModificationDate.toLocaleDateString("pl")}
    </Typography>
  );

  return (
    <Stack
      data-testid="comment-genuine"
      component="article"
      direction={orientation}
      spacing={3}
      padding={2}
      borderRadius={10}
      alignItems={orientation === "column" ? "center" : "start"}
      border={(theme) => `dotted 2px ${theme.palette.divider}`}
    >
      <Stack
        spacing={2}
        maxWidth={60}
        justifyContent="center"
        alignItems="center"
      >
        <Avatar
          variant="solid"
          color="warning"
          src={userAvatar}
          alt={`${userName}'s avatar`}
        />
        <Typography variant="solid" fontFamily="monospace">
          {userName}
        </Typography>
        {orientation === "column" && dateComponent}
      </Stack>
      <Stack flex={1}>
        <Typography level="h6">{title}</Typography>
        <Typography variant="soft" flex={1}>
          {description}
        </Typography>
      </Stack>
      {orientation === "row" && dateComponent}
    </Stack>
  );
};
