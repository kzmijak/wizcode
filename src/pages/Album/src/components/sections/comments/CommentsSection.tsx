import { Divider, Typography } from "@mui/joy";
import { FC } from "react";
import { AlbumComment, AlbumCommentProps } from "./AlbumComment";
import {
  albumCommentsState,
  useAlbumCommentsFetch,
} from "modules/AlbumCommentsState";
import { useRecoilValue } from "recoil";
import { AlbumModel } from "models/Album";
import { AlbumCommentSkeleton } from "./AlbumCommentSkeleton";

type CommentsSectionProps = Pick<AlbumModel, "id"> &
  Pick<AlbumCommentProps, "orientation">;

export const CommentsSection: FC<CommentsSectionProps> = ({
  orientation,
  ...album
}) => {
  const { status } = useAlbumCommentsFetch(album.id);
  const albumComments = useRecoilValue(albumCommentsState(album.id));

  const skeleton = [...Array(8).keys()].map((key) => (
    <AlbumCommentSkeleton key={key} orientation={orientation} />
  ));
  const comments = albumComments.map((comment) => (
    <AlbumComment key={comment.id} {...comment} orientation={orientation} />
  ));

  return (
    <>
      <Typography level="h4">Comments</Typography>
      <Divider />
      {status === "loading" ? skeleton : comments}
    </>
  );
};
