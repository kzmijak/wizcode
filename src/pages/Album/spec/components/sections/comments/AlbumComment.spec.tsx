import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { AlbumComment } from "../../../../src/components/sections/comments/AlbumComment";

describe("AlbumComment", () => {
  const Component = () => (
    <AlbumComment
      description="Comment description"
      lastModificationDate={new Date("06.19.2023")}
      title="Comment title"
      userAvatar="imgsrc"
      userName="User name"
    />
  );

  it("should match snapshot", () => {
    expect(render(<Component />)).toMatchSnapshot();
  });

  it("should render static component", () => {
    render(<Component />);

    expect(screen.queryByText("Comment description")).toBeInTheDocument();
    expect(screen.queryByText("Comment title")).toBeInTheDocument();
    expect(screen.queryByText("User name")).toBeInTheDocument();
    expect(screen.queryByText("19.06.2023")).toBeInTheDocument();
  });

  it("should properly render image", () => {
    render(<Component />);

    expect(screen.queryByRole("img")).toBeInTheDocument();
    expect(screen.queryByRole("img")).toHaveAttribute(
      "alt",
      "User name's avatar"
    );
  });
});
