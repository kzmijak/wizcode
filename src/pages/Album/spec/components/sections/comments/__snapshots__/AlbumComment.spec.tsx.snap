// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AlbumComment > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="MuiStack-root css-kg3pic-JoyStack-root"
        data-testid="comment-genuine"
      >
        <div
          class="MuiStack-root css-8e38t1-JoyStack-root"
        >
          <div
            class="MuiAvatar-root MuiAvatar-variantSolid MuiAvatar-colorWarning MuiAvatar-sizeMd css-1w8mgdh-JoyAvatar-root"
          >
            <img
              alt="User name's avatar"
              class="MuiAvatar-img css-9ae35a-JoyAvatar-img"
              src="imgsrc"
            />
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorNeutral MuiTypography-variantSolid css-g7l4sp-JoyTypography-root"
          >
            User name
          </p>
        </div>
        <div
          class="MuiStack-root css-1j7mj4l-JoyStack-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-7ui8qe-JoyTypography-root"
          >
            Comment title
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorNeutral MuiTypography-variantSoft css-2sqj6q-JoyTypography-root"
          >
            Comment description
          </p>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 css-c4auoj-JoyTypography-root"
        >
          19.06.2023
        </p>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="MuiStack-root css-kg3pic-JoyStack-root"
      data-testid="comment-genuine"
    >
      <div
        class="MuiStack-root css-8e38t1-JoyStack-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-variantSolid MuiAvatar-colorWarning MuiAvatar-sizeMd css-1w8mgdh-JoyAvatar-root"
        >
          <img
            alt="User name's avatar"
            class="MuiAvatar-img css-9ae35a-JoyAvatar-img"
            src="imgsrc"
          />
        </div>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-colorNeutral MuiTypography-variantSolid css-g7l4sp-JoyTypography-root"
        >
          User name
        </p>
      </div>
      <div
        class="MuiStack-root css-1j7mj4l-JoyStack-root"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-7ui8qe-JoyTypography-root"
        >
          Comment title
        </h6>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-colorNeutral MuiTypography-variantSoft css-2sqj6q-JoyTypography-root"
        >
          Comment description
        </p>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 css-c4auoj-JoyTypography-root"
      >
        19.06.2023
      </p>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
